File: C/Gemini/NBodySimulation.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <time.h>

// Define the structure for a body
typedef struct {
    double mass;
    double position[3];
    double velocity[3];
} Body;

// Define the structure for the system
typedef struct {
    int num_bodies;
    Body *bodies;
} System;

// Gravitational constant
const double G = 6.67430e-11;

// Function to calculate the distance squared between two bodies
double distance_squared(const Body *b1, const Body *b2) {
    double dx = b1->position[0] - b2->position[0];
    double dy = b1->position[1] - b2->position[1];
    double dz = b1->position[2] - b2->position[2];
    return dx * dx + dy * dy + dz * dz;
}

// Function to calculate the force on body i due to body j
void calculate_force(const Body *body_i, const Body *body_j, double force[3]) {
    double r2 = distance_squared(body_i, body_j);
    if (r2 > 1e-9) { // Avoid division by zero for very close bodies
        double r = sqrt(r2);
        double magnitude = (G * body_i->mass * body_j->mass) / r2;
        force[0] = magnitude * (body_j->position[0] - body_i->position[0]) / r;
        force[1] = magnitude * (body_j->position[1] - body_i->position[1]) / r;
        force[2] = magnitude * (body_j->position[2] - body_i->position[2]) / r;
    } else {
        force[0] = force[1] = force[2] = 0.0;
    }
}

// Function to calculate the total energy of the system
// double calculate_total_energy(const System *system) {
//     double kinetic_energy = 0.0;
//     double potential_energy = 0.0;

//     for (int i = 0; i < system->num_bodies; i++) {
//         // Kinetic energy
//         kinetic_energy += 0.5 * system->bodies[i].mass * (system->bodies[i].velocity[0] * system->bodies[i].velocity[0] +
//                                                           system->bodies[i].velocity[1] * system->bodies[i].velocity[1] +
//                                                           system->bodies[i].velocity[2] * system->bodies[i].velocity[2]);

//         // Potential energy
//         for (int j = i + 1; j < system->num_bodies; j++) {
//             potential_energy -= (G * system->bodies[i].mass * system->bodies[j].mass) / sqrt(distance_squared(&system->bodies[i], &system->bodies[j]));
//         }
//     }

//     return kinetic_energy + potential_energy;
// }

// Function to initialize a system with a central body and orbiting smaller bodies
System* initialize_circular_orbits(int num_orbiting_bodies, double central_mass, double orbit_radius, double orbiting_mass) {
    System *system = (System*)malloc(sizeof(System));
    if (system == NULL) {
        fprintf(stderr, "Memory allocation failed for system.\n");
        exit(EXIT_FAILURE);
    }
    system->num_bodies = num_orbiting_bodies + 1;
    system->bodies = (Body*)malloc(sizeof(Body) * system->num_bodies);
    if (system->bodies == NULL) {
        fprintf(stderr, "Memory allocation failed for bodies.\n");
        free(system);
        exit(EXIT_FAILURE);
    }

    // Initialize the central body
    system->bodies[0].mass = central_mass;
    system->bodies[0].position[0] = 0.0;
    system->bodies[0].position[1] = 0.0;
    system->bodies[0].position[2] = 0.0;
    system->bodies[0].velocity[0] = 0.0;
    system->bodies[0].velocity[1] = 0.0;
    system->bodies[0].velocity[2] = 0.0;

    // Initialize the orbiting bodies
    for (int i = 1; i < system->num_bodies; i++) {
        system->bodies[i].mass = orbiting_mass;
        system->bodies[i].position[0] = orbit_radius * cos(2 * M_PI * (i - 1) / num_orbiting_bodies);
        system->bodies[i].position[1] = orbit_radius * sin(2 * M_PI * (i - 1) / num_orbiting_bodies);
        system->bodies[i].position[2] = 0.0; // Orbiting in the xy-plane

        // Calculate the orbital velocity for a circular orbit
        double orbital_speed = sqrt(G * central_mass / orbit_radius);
        system->bodies[i].velocity[0] = -orbital_speed * sin(2 * M_PI * (i - 1) / num_orbiting_bodies);
        system->bodies[i].velocity[1] = orbital_speed * cos(2 * M_PI * (i - 1) / num_orbiting_bodies);
        system->bodies[i].velocity[2] = 0.0;
    }

    return system;
}

// Function to perform one kick-step
void kick_step(System *system, double dt) {
    int n = system->num_bodies;
    double (*forces)[3] = malloc(sizeof(double[n][3]));
    if (forces == NULL) {
        fprintf(stderr, "Memory allocation failed for forces.\n");
        exit(EXIT_FAILURE);
    }

    // Calculate forces
    for (int i = 0; i < n; i++) {
        forces[i][0] = forces[i][1] = forces[i][2] = 0.0;
        for (int j = 0; j < n; j++) {
            if (i != j) {
                double f[3];
                calculate_force(&system->bodies[i], &system->bodies[j], f);
                forces[i][0] += f[0];
                forces[i][1] += f[1];
                forces[i][2] += f[2];
            }
        }
    }

    // Update velocities (kick)
    for (int i = 0; i < n; i++) {
        system->bodies[i].velocity[0] += (forces[i][0] / system->bodies[i].mass) * dt * 0.5;
        system->bodies[i].velocity[1] += (forces[i][1] / system->bodies[i].mass) * dt * 0.5;
        system->bodies[i].velocity[2] += (forces[i][2] / system->bodies[i].mass) * dt * 0.5;
    }

    free(forces);
}

// Function to perform one drift-step
void drift_step(System *system, double dt) {
    for (int i = 0; i < system->num_bodies; i++) {
        system->bodies[i].position[0] += system->bodies[i].velocity[0] * dt;
        system->bodies[i].position[1] += system->bodies[i].velocity[1] * dt;
        system->bodies[i].position[2] += system->bodies[i].velocity[2] * dt;
    }
}

// Function to perform one full first-order kick-step
void first_order_kick_step(System *system, double dt) {
    kick_step(system, dt);
    drift_step(system, dt);
    kick_step(system, dt);
}

// Function to free the memory allocated for the system
void free_system(System *system) {
    if (system != NULL) {
        free(system->bodies);
        free(system);
    }
}

int main() {
    int num_orbiting_bodies = 10000;
    double central_mass = 1.989e30; // Mass of the Sun (kg)
    double orbit_radius = 1.496e11; // 1 AU (m)
    double orbiting_mass = 5.972e24 / num_orbiting_bodies; // Mass of the Earth (kg)
    int num_steps = 100;
    double time_step = 3600 * 24 * 7; // 1 week in seconds

    // Initialize the system
    System *system = initialize_circular_orbits(num_orbiting_bodies, central_mass, orbit_radius, orbiting_mass);
    if (system == NULL) {
        return 1;
    }

    printf("Initial number of bodies: %d\n", system->num_bodies);

    // Calculate initial energy
    // double initial_energy = calculate_total_energy(system);
    // printf("Initial total energy: %e J\n", initial_energy);

    // Run the simulation
    printf("Running simulation for %d steps...\n", num_steps);
    time_t start_time = time(NULL);
    for (int step = 0; step < num_steps; step++) {
        first_order_kick_step(system, time_step);
        if ((step + 1) % 100 == 0) {
            printf("Step %d completed.\n", step + 1);
        }
    }
    time_t end_time = time(NULL);
    double elapsed_time = difftime(end_time, start_time);
    printf("Simulation finished in %.2f seconds.\n", elapsed_time);

    // Calculate final energy
    // double final_energy = calculate_total_energy(system);
    // printf("Final total energy: %e J\n", final_energy);

    // // Calculate the energy difference
    // double energy_difference = fabs(final_energy - initial_energy);
    // double relative_energy_difference = energy_difference / fabs(initial_energy);
    // printf("Absolute energy difference: %e J\n", energy_difference);
    // printf("Relative energy difference: %e\n", relative_energy_difference);

    // Free the memory
    free_system(system);

    return 0;
}

File: Go/Gemini/NBodySimulation.go
package main

import (
	"fmt"
	"math"
	"math/rand"
	"time"
)

// Body represents a celestial body with mass, position, and velocity.
type Body struct {
	Mass     float64
	Position [3]float64
	Velocity [3]float64
}

// System represents a collection of celestial bodies.
type System struct {
	Bodies []Body
}

// GravitationalConstant is the universal gravitational constant.
const GravitationalConstant = 6.67430e-11

// distanceSquared calculates the squared Euclidean distance between two bodies.
func distanceSquared(b1, b2 *Body) float64 {
	dx := b1.Position[0] - b2.Position[0]
	dy := b1.Position[1] - b2.Position[1]
	dz := b1.Position[2] - b2.Position[2]
	return dx*dx + dy*dy + dz*dz
}

// calculateForce calculates the gravitational force on body i due to body j.
func calculateForce(bodyI, bodyJ *Body, force *[3]float64) {
	rSq := distanceSquared(bodyI, bodyJ)
	if rSq > 1e-9 { // Avoid division by zero for very close bodies
		r := math.Sqrt(rSq)
		magnitude := (GravitationalConstant * bodyI.Mass * bodyJ.Mass) / rSq
		force[0] += magnitude * (bodyJ.Position[0] - bodyI.Position[0]) / r
		force[1] += magnitude * (bodyJ.Position[1] - bodyI.Position[1]) / r
		force[2] += magnitude * (bodyJ.Position[2] - bodyI.Position[2]) / r
	}
}

// calculateTotalEnergy calculates the total energy of the system.
// func calculateTotalEnergy(s *System) float64 {
// 	kineticEnergy := 0.0
// 	potentialEnergy := 0.0

// 	for i := range s.Bodies {
// 		// Kinetic energy
// 		vSq := s.Bodies[i].Velocity[0]*s.Bodies[i].Velocity[0] +
// 			s.Bodies[i].Velocity[1]*s.Bodies[i].Velocity[1] +
// 			s.Bodies[i].Velocity[2]*s.Bodies[i].Velocity[2]
// 		kineticEnergy += 0.5 * s.Bodies[i].Mass * vSq

// 		// Potential energy
// 		for j := i + 1; j < len(s.Bodies); j++ {
// 			r := math.Sqrt(distanceSquared(&s.Bodies[i], &s.Bodies[j]))
// 			potentialEnergy -= (GravitationalConstant * s.Bodies[i].Mass * s.Bodies[j].Mass) / r
// 		}
// 	}

// 	return kineticEnergy + potentialEnergy
// }

// initializeCircularOrbits initializes a system with a central body and orbiting smaller bodies.
func initializeCircularOrbits(numOrbiting int, centralMass, orbitRadius, orbitingMass float64) *System {
	system := &System{
		Bodies: make([]Body, numOrbiting+1),
	}

	// Initialize the central body
	system.Bodies[0] = Body{
		Mass:     centralMass,
		Position: [3]float64{0, 0, 0},
		Velocity: [3]float64{0, 0, 0},
	}

	// Initialize the orbiting bodies
	for i := 0; i < numOrbiting; i++ {
		angle := 2 * math.Pi * float64(i) / float64(numOrbiting)
		system.Bodies[i+1] = Body{
			Mass: orbitingMass,
			Position: [3]float64{
				orbitRadius * math.Cos(angle),
				orbitRadius * math.Sin(angle),
				0,
			},
		}

		// Calculate the orbital velocity for a circular orbit
		orbitalSpeed := math.Sqrt(GravitationalConstant * centralMass / orbitRadius)
		system.Bodies[i+1].Velocity = [3]float64{
			-orbitalSpeed * math.Sin(angle),
			orbitalSpeed * math.Cos(angle),
			0,
		}
	}

	return system
}

// kickStep performs the "kick" part of the first-order kick-step method.
func kickStep(s *System, dt float64) {
	numBodies := len(s.Bodies)
	forces := make([][3]float64, numBodies)

	// Calculate forces
	for i := 0; i < numBodies; i++ {
		for j := 0; j < numBodies; j++ {
			if i != j {
				calculateForce(&s.Bodies[i], &s.Bodies[j], &forces[i])
			}
		}
	}

	// Update velocities
	for i := 0; i < numBodies; i++ {
		s.Bodies[i].Velocity[0] += (forces[i][0] / s.Bodies[i].Mass) * dt * 0.5
		s.Bodies[i].Velocity[1] += (forces[i][1] / s.Bodies[i].Mass) * dt * 0.5
		s.Bodies[i].Velocity[2] += (forces[i][2] / s.Bodies[i].Mass) * dt * 0.5
	}
}

// driftStep performs the "drift" part of the first-order kick-step method.
func driftStep(s *System, dt float64) {
	for i := range s.Bodies {
		s.Bodies[i].Position[0] += s.Bodies[i].Velocity[0] * dt
		s.Bodies[i].Position[1] += s.Bodies[i].Velocity[1] * dt
		s.Bodies[i].Position[2] += s.Bodies[i].Velocity[2] * dt
	}
}

// firstOrderKickStep performs one full first-order kick-step.
func firstOrderKickStep(s *System, dt float64) {
	kickStep(s, dt)
	driftStep(s, dt)
	kickStep(s, dt)
}

func main() {
	numOrbitingBodies := 10000
	centralMass := 1.989e30  // Mass of the Sun (kg)
	orbitRadius := 1.496e11  // 1 AU (m)
	orbitingMass := 5.972e24 // Mass of the Earth (kg)
	numSteps := 100
	timeStep := 3600.0 * 24.0 * 7.0 // 1 week in seconds

	rand.Seed(time.Now().UnixNano())

	// Initialize the system
	system := initializeCircularOrbits(numOrbitingBodies, centralMass, orbitRadius, orbitingMass)
	fmt.Printf("Initial number of bodies: %d\n", len(system.Bodies))

	// Calculate initial energy
	// initialEnergy := calculateTotalEnergy(system)
	// fmt.Printf("Initial total energy: %e J\n", initialEnergy)

	// Run the simulation
	fmt.Printf("Running simulation for %d steps...\n", numSteps)
	startTime := time.Now()
	for step := 0; step < numSteps; step++ {
		firstOrderKickStep(system, timeStep)
		if (step+1)%100 == 0 {
			fmt.Printf("Step %d completed.\n", step+1)
		}
	}
	elapsedTime := time.Since(startTime)
	fmt.Printf("Simulation finished in %s.\n", elapsedTime)

	// Calculate final energy
	// finalEnergy := calculateTotalEnergy(system)
	// fmt.Printf("Final total energy: %e J\n", finalEnergy)

	// // Calculate the energy difference
	// energyDifference := math.Abs(finalEnergy - initialEnergy)
	// relativeEnergyDifference := energyDifference / math.Abs(initialEnergy)
	// fmt.Printf("Absolute energy difference: %e J\n", energyDifference)
	// fmt.Printf("Relative energy difference: %e\n", relativeEnergyDifference)
	fmt.Printf("Body[0] %e %e %e", system.Bodies[0].Position[0], system.Bodies[0].Position[1], system.Bodies[0].Position[2])

}
